src/sys/windows/afd.rs

```diff
@@ -0,0 +1,184 @@
+use std::ffi::OsStr;
+use std::fmt;
+use std::fs::File;
+use std::io;
+use std::mem::size_of;
+use std::ptr::null_mut;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use miow::iocp::CompletionPort;
+
+use std::os::windows::ffi::OsStrExt;
+use std::os::windows::io::{AsRawHandle, FromRawHandle, RawHandle};
+
+use winapi::shared::ntdef::{
+    HANDLE, LARGE_INTEGER, NTSTATUS, OBJECT_ATTRIBUTES, PVOID, ULONG, UNICODE_STRING,
+};
+use winapi::shared::ntstatus::{STATUS_NOT_FOUND, STATUS_PENDING, STATUS_SUCCESS};
+use winapi::um::handleapi::INVALID_HANDLE_VALUE;
+use winapi::um::winbase::{SetFileCompletionNotificationModes, FILE_SKIP_SET_EVENT_ON_HANDLE};
+use winapi::um::winnt::SYNCHRONIZE;
+use winapi::um::winnt::{FILE_SHARE_READ, FILE_SHARE_WRITE};
+
+use ntapi::ntioapi::FILE_OPEN;
+use ntapi::ntioapi::IO_STATUS_BLOCK;
+use ntapi::ntioapi::{NtCancelIoFileEx, NtCreateFile, NtDeviceIoControlFile};
+use ntapi::ntrtl::RtlNtStatusToDosError;
+
+const IOCTL_AFD_POLL: ULONG = 0x00012024;
+const AFD_HELPER_NAME: &'static str = "\\Device\\Afd\\Mio";
+
+static NEXT_TOKEN: AtomicUsize = AtomicUsize::new(0);
+
+#[derive(Debug)]
+pub struct Afd {
+    fd: File,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct AfdPollHandleInfo {
+    pub handle: HANDLE,
+    pub events: ULONG,
+    pub status: NTSTATUS,
+}
+
+unsafe impl Send for AfdPollHandleInfo {}
+unsafe impl Sync for AfdPollHandleInfo {}
+
+#[repr(C)]
+pub struct AfdPollInfo {
+    pub timeout: LARGE_INTEGER,
+    // Can have only value 1.
+    pub number_of_handles: ULONG,
+    pub exclusive: ULONG,
+    pub handles: [AfdPollHandleInfo; 1],
+}
+
+impl fmt::Debug for AfdPollInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AfdPollInfo").finish()
+    }
+}
+
+impl Afd {
+    pub fn new(cp: &CompletionPort) -> io::Result<Afd> {
+        let mut afd_helper_name = OsStr::new(AFD_HELPER_NAME)
```

Comment:
It's unfortunate that this is done every time a handle to the afd driver is created.
It'd be nicer to make afd_helper_attributes a static global, maybe with the help of lazy_static.

Reply:
But doing so in a follow up patch would be okay.

Reply:
Done.